google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -101,23 +104,35 @@ protected void doStart() {
   @Override
   protected void doStop() {
     messageDispatcher.stop();
-    requestObserver.onError(Status.CANCELLED.asException());
-    notifyStopped();
+
+    lock.lock();
+    try {
+      requestObserver.onError(Status.CANCELLED.asException());
+    } finally {
+      lock.unlock();
+      notifyStopped();
+    }
   }
 
   private class StreamingPullResponseObserver
       implements ClientResponseObserver<StreamingPullRequest, StreamingPullResponse> {
 
     final SettableFuture<Void> errorFuture;
+    ClientCallStreamObserver<StreamingPullRequest> thisRequestObserver;
 
     StreamingPullResponseObserver(SettableFuture<Void> errorFuture) {
       this.errorFuture = errorFuture;
     }
 
     @Override
     public void beforeStart(ClientCallStreamObserver<StreamingPullRequest> requestObserver) {
-      StreamingSubscriberConnection.this.requestObserver = requestObserver;
-      requestObserver.disableAutoInboundFlowControl();
+      thisRequestObserver = requestObserver;
+      lock.lock();
```

Comment:
This should not require a lock, we have the guarantee here that nothing could have killed the stream since this is beforeStart, we cannot have had our error callback hit. We can also just use thisRequestObserver since it does not need to be lock protected.

Also why the change from this.requestObserver to thisRequestObserver?

Reply:
You're right about the lock. Done.

I documented `thisRequestObserver` in the `onNext` method. I moved the doc here so it reads better.